File: sdk/ts-agent-sdk/src/codec.ts
/**
 * Wire format codec matching Rust kernel-core codec.rs byte-for-byte.
 *
 * CRITICAL: All integers are LITTLE-ENDIAN to match the Rust implementation.
 * This is consensus-critical -- any deviation will produce different hashes.
 */

import type { ActionV1, AgentOutput, KernelInputV1, KernelJournalV1 } from "./types.js";
import {
  ExecutionStatus,
  JOURNAL_SIZE,
  MAX_ACTIONS_PER_OUTPUT,
  MAX_ACTION_PAYLOAD_BYTES,
  MAX_AGENT_INPUT_BYTES,
  MAX_AGENT_OUTPUT_BYTES,
  MAX_SINGLE_ACTION_BYTES,
  PROTOCOL_VERSION,
  KERNEL_VERSION,
} from "./types.js";

// ============================================================================
// Helper Functions - Writing (Little-Endian)
// ============================================================================

/** Write a u32 as little-endian into a DataView at the given offset. Returns new offset. */
export function writeU32LE(view: DataView, offset: number, value: number): number {
  view.setUint32(offset, value, true);
  return offset + 4;
}

/** Write a u64 as little-endian into a DataView at the given offset. Returns new offset. */
export function writeU64LE(view: DataView, offset: number, value: bigint): number {
  view.setBigUint64(offset, value, true);
  return offset + 8;
}

/** Write a Uint8Array into a buffer at the given offset. Returns new offset. */
export function writeBytes(buf: Uint8Array, offset: number, data: Uint8Array): number {
  buf.set(data, offset);
  return offset + data.length;
}

// ============================================================================
// Helper Functions - Reading (Little-Endian)
// ============================================================================

/** Read a u32 from little-endian bytes at offset. Returns [value, newOffset]. */
export function readU32LE(view: DataView, offset: number): [number, number] {
  if (offset + 4 > view.byteLength) {
    throw new RangeError(`readU32LE: unexpected end of input at offset ${offset}`);
  }
  return [view.getUint32(offset, true), offset + 4];
}

/** Read a u64 from little-endian bytes at offset. Returns [value, newOffset]. */
export function readU64LE(view: DataView, offset: number): [bigint, number] {
  if (offset + 8 > view.byteLength) {
    throw new RangeError(`readU64LE: unexpected end of input at offset ${offset}`);
  }
  return [view.getBigUint64(offset, true), offset + 8];
}

/** Read a u8 at offset. Returns [value, newOffset]. */
export function readU8(data: Uint8Array, offset: number): [number, number] {
  if (offset >= data.length) {
    throw new RangeError(`readU8: unexpected end of input at offset ${offset}`);
  }
  return [data[offset]!, offset + 1];
}

/** Read N bytes at offset, returns a copy. Returns [bytes, newOffset]. */
export function readBytes(data: Uint8Array, offset: number, length: number): [Uint8Array, number] {
  const end = offset + length;
  if (end > data.length) {
    throw new RangeError(`readBytes: unexpected end of input at offset ${offset}, need ${length} bytes`);
  }
  return [data.slice(offset, end), end];
}

/** Read exactly 32 bytes at offset, returns a copy. */
export function readBytes32(data: Uint8Array, offset: number): [Uint8Array, number] {
  return readBytes(data, offset, 32);
}

// ============================================================================
// ActionV1 Codec
// ============================================================================

/**
 * Compute the encoded length of a single ActionV1.
 *
 * Layout: action_type (4) + target (32) + payload_len (4) + payload
 */
function actionEncodedLen(action: ActionV1): number {
  return 40 + action.payload.length;
}

/**
 * Encode a single ActionV1 into a buffer at the given offset.
 * Returns the new offset after writing.
 */
function encodeActionInto(buf: Uint8Array, view: DataView, offset: number, action: ActionV1): number {
  if (action.payload.length > MAX_ACTION_PAYLOAD_BYTES) {
    throw new Error(
      `Action payload too large: ${action.payload.length} bytes (limit: ${MAX_ACTION_PAYLOAD_BYTES})`
    );
  }

  offset = writeU32LE(view, offset, action.actionType);
  offset = writeBytes(buf, offset, action.target);
  offset = writeU32LE(view, offset, action.payload.length);
  offset = writeBytes(buf, offset, action.payload);
  return offset;
}

/**
 * Decode a single ActionV1 from a buffer slice.
 * The slice must contain exactly one action (no trailing bytes).
 */
function decodeAction(data: Uint8Array): ActionV1 {
  if (data.length < 40) {
    throw new Error("ActionV1: unexpected end of input");
  }

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let offset = 0;

  let actionType: number;
  [actionType, offset] = readU32LE(view, offset);

  let target: Uint8Array;
  [target, offset] = readBytes32(data, offset);

  let payloadLen: number;
  [payloadLen, offset] = readU32LE(view, offset);

  if (payloadLen > MAX_ACTION_PAYLOAD_BYTES) {
    throw new Error(
      `Action payload too large: ${payloadLen} bytes (limit: ${MAX_ACTION_PAYLOAD_BYTES})`
    );
  }

  let payload: Uint8Array;
  [payload, offset] = readBytes(data, offset, payloadLen);

  if (offset !== data.length) {
    throw new Error("ActionV1: trailing bytes after action");
  }

  return { actionType, target, payload };
}

// ============================================================================
// AgentOutput Codec
// ============================================================================

/**
 * Encode an AgentOutput into its canonical wire format.
 *
 * Layout (little-endian):
 * - action_count: u32 (4 bytes)
 * - for each action:
 *   - action_len: u32 (4 bytes)
 *   - ActionV1 encoding (variable)
 *
 * Action order is preserved as-is (agent is responsible for deterministic ordering).
 */
export function encodeAgentOutput(output: AgentOutput): Uint8Array {
  const n = output.actions.length;
  if (n > MAX_ACTIONS_PER_OUTPUT) {
    throw new Error(`Too many actions: ${n} (limit: ${MAX_ACTIONS_PER_OUTPUT})`);
  }

  // Compute total size
  let totalSize = 4; // action_count
  for (const action of output.actions) {
    const actionLen = actionEncodedLen(action);
    if (actionLen > MAX_SINGLE_ACTION_BYTES) {
      throw new Error(`Action too large: ${actionLen} bytes (limit: ${MAX_SINGLE_ACTION_BYTES})`);
    }
    totalSize += 4 + actionLen; // length prefix + action data
  }

  if (totalSize > MAX_AGENT_OUTPUT_BYTES) {
    throw new Error(`Output too large: ${totalSize} bytes (limit: ${MAX_AGENT_OUTPUT_BYTES})`);
  }

  const buf = new Uint8Array(totalSize);
  const view = new DataView(buf.buffer);
  let offset = 0;

  offset = writeU32LE(view, offset, n);

  for (const action of output.actions) {
    const actionLen = actionEncodedLen(action);
    offset = writeU32LE(view, offset, actionLen);
    offset = encodeActionInto(buf, view, offset, action);
  }

  return buf;
}

/**
 * Decode an AgentOutput from its canonical wire format.
 */
export function decodeAgentOutput(data: Uint8Array): AgentOutput {
  if (data.length > MAX_AGENT_OUTPUT_BYTES) {
    throw new Error(`Output too large: ${data.length} bytes (limit: ${MAX_AGENT_OUTPUT_BYTES})`);
  }

  if (data.length < 4) {
    throw new Error("AgentOutput: unexpected end of input");
  }

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let offset = 0;

  let actionCount: number;
  [actionCount, offset] = readU32LE(view, offset);

  if (actionCount > MAX_ACTIONS_PER_OUTPUT) {
    throw new Error(`Too many actions: ${actionCount} (limit: ${MAX_ACTIONS_PER_OUTPUT})`);
  }

  const actions: ActionV1[] = [];

  for (let i = 0; i < actionCount; i++) {
    let actionLen: number;
    [actionLen, offset] = readU32LE(view, offset);

    if (actionLen > MAX_SINGLE_ACTION_BYTES) {
      throw new Error(`Action too large: ${actionLen} bytes (limit: ${MAX_SINGLE_ACTION_BYTES})`);
    }

    const actionEnd = offset + actionLen;
    if (actionEnd > data.length) {
      throw new Error("AgentOutput: unexpected end of input while reading action");
    }

    const actionData = data.slice(offset, actionEnd);
    actions.push(decodeAction(actionData));
    offset = actionEnd;
  }

  if (offset !== data.length) {
    throw new Error("AgentOutput: trailing bytes after output");
  }

  return { actions };
}

// ============================================================================
// KernelInputV1 Codec
// ============================================================================

/**
 * Decode a KernelInputV1 from its canonical wire format.
 *
 * Layout (little-endian):
 * - protocol_version: u32 (4)
 * - kernel_version: u32 (4)
 * - agent_id: bytes32 (32)
 * - agent_code_hash: bytes32 (32)
 * - constraint_set_hash: bytes32 (32)
 * - input_root: bytes32 (32)
 * - execution_nonce: u64 (8)
 * - opaque_agent_inputs_len: u32 (4)
 * - opaque_agent_inputs: variable
 *
 * Minimum size: 148 bytes (with empty opaque_agent_inputs).
 */
export function decodeKernelInput(data: Uint8Array): KernelInputV1 {
  if (data.length < 148) {
    throw new Error("KernelInputV1: unexpected end of input");
  }

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let offset = 0;

  let protocolVersion: number;
  [protocolVersion, offset] = readU32LE(view, offset);

  let kernelVersion: number;
  [kernelVersion, offset] = readU32LE(view, offset);

  let agentId: Uint8Array;
  [agentId, offset] = readBytes32(data, offset);

  let agentCodeHash: Uint8Array;
  [agentCodeHash, offset] = readBytes32(data, offset);

  let constraintSetHash: Uint8Array;
  [constraintSetHash, offset] = readBytes32(data, offset);

  let inputRoot: Uint8Array;
  [inputRoot, offset] = readBytes32(data, offset);

  let executionNonce: bigint;
  [executionNonce, offset] = readU64LE(view, offset);

  let inputsLen: number;
  [inputsLen, offset] = readU32LE(view, offset);

  if (inputsLen > MAX_AGENT_INPUT_BYTES) {
    throw new Error(`Agent inputs too large: ${inputsLen} bytes (limit: ${MAX_AGENT_INPUT_BYTES})`);
  }

  let opaqueAgentInputs: Uint8Array;
  [opaqueAgentInputs, offset] = readBytes(data, offset, inputsLen);

  if (offset !== data.length) {
    throw new Error("KernelInputV1: trailing bytes");
  }

  return {
    protocolVersion,
    kernelVersion,
    agentId,
    agentCodeHash,
    constraintSetHash,
    inputRoot,
    executionNonce,
    opaqueAgentInputs,
  };
}

/**
 * Encode a KernelInputV1 into its canonical wire format.
 */
export function encodeKernelInput(input: KernelInputV1): Uint8Array {
  const inputsLen = input.opaqueAgentInputs.length;
  if (inputsLen > MAX_AGENT_INPUT_BYTES) {
    throw new Error(`Agent inputs too large: ${inputsLen} bytes (limit: ${MAX_AGENT_INPUT_BYTES})`);
  }

  const totalSize = 148 + inputsLen;
  const buf = new Uint8Array(totalSize);
  const view = new DataView(buf.buffer);
  let offset = 0;

  offset = writeU32LE(view, offset, input.protocolVersion);
  offset = writeU32LE(view, offset, input.kernelVersion);
  offset = writeBytes(buf, offset, input.agentId);
  offset = writeBytes(buf, offset, input.agentCodeHash);
  offset = writeBytes(buf, offset, input.constraintSetHash);
  offset = writeBytes(buf, offset, input.inputRoot);
  offset = writeU64LE(view, offset, input.executionNonce);
  offset = writeU32LE(view, offset, inputsLen);
  offset = writeBytes(buf, offset, input.opaqueAgentInputs);

  return buf;
}

// ============================================================================
// KernelJournalV1 Codec
// ============================================================================

/**
 * Encode a KernelJournalV1 into its canonical wire format.
 *
 * Always produces exactly 209 bytes.
 */
export function encodeKernelJournal(journal: KernelJournalV1): Uint8Array {
  const buf = new Uint8Array(JOURNAL_SIZE);
  const view = new DataView(buf.buffer);
  let offset = 0;

  offset = writeU32LE(view, offset, journal.protocolVersion);
  offset = writeU32LE(view, offset, journal.kernelVersion);
  offset = writeBytes(buf, offset, journal.agentId);
  offset = writeBytes(buf, offset, journal.agentCodeHash);
  offset = writeBytes(buf, offset, journal.constraintSetHash);
  offset = writeBytes(buf, offset, journal.inputRoot);
  offset = writeU64LE(view, offset, journal.executionNonce);
  offset = writeBytes(buf, offset, journal.inputCommitment);
  offset = writeBytes(buf, offset, journal.actionCommitment);

  // ExecutionStatus: Success = 0x01, Failure = 0x02
  buf[offset] = journal.executionStatus;

  return buf;
}

/**
 * Decode a KernelJournalV1 from its canonical wire format.
 *
 * Input must be exactly 209 bytes.
 */
export function decodeKernelJournal(data: Uint8Array): KernelJournalV1 {
  if (data.length !== JOURNAL_SIZE) {
    throw new Error(
      `KernelJournalV1: invalid length ${data.length}, expected ${JOURNAL_SIZE}`
    );
  }

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let offset = 0;

  let protocolVersion: number;
  [protocolVersion, offset] = readU32LE(view, offset);

  let kernelVersion: number;
  [kernelVersion, offset] = readU32LE(view, offset);

  let agentId: Uint8Array;
  [agentId, offset] = readBytes32(data, offset);

  let agentCodeHash: Uint8Array;
  [agentCodeHash, offset] = readBytes32(data, offset);

  let constraintSetHash: Uint8Array;
  [constraintSetHash, offset] = readBytes32(data, offset);

  let inputRoot: Uint8Array;
  [inputRoot, offset] = readBytes32(data, offset);

  let executionNonce: bigint;
  [executionNonce, offset] = readU64LE(view, offset);

  let inputCommitment: Uint8Array;
  [inputCommitment, offset] = readBytes32(data, offset);

  let actionCommitment: Uint8Array;
  [actionCommitment, offset] = readBytes32(data, offset);

  const statusByte = data[offset]!;
  if (statusByte !== ExecutionStatus.Success && statusByte !== ExecutionStatus.Failure) {
    throw new Error(`Invalid execution status byte: 0x${statusByte.toString(16).padStart(2, "0")}`);
  }

  return {
    protocolVersion,
    kernelVersion,
    agentId,
    agentCodeHash,
    constraintSetHash,
    inputRoot,
    executionNonce,
    inputCommitment,
    actionCommitment,
    executionStatus: statusByte as ExecutionStatus,
  };
}

File: sdk/py-agent-sdk/tokagent/codec.py
"""Wire-format codec matching the Rust canonical codec byte-for-byte.

All multi-byte integers use **little-endian** encoding (matching Rust's
``to_le_bytes``).  The only exception is ABI-encoded payloads inside
actions, which follow Solidity's big-endian uint256 convention -- but
those are opaque bytes from the codec's perspective.

Wire layouts
============

KernelInputV1 (variable length, min 148 bytes)::

    protocol_version   u32 LE   4
    kernel_version     u32 LE   4
    agent_id           bytes32  32
    agent_code_hash    bytes32  32
    constraint_set_hash bytes32 32
    input_root         bytes32  32
    execution_nonce    u64 LE   8
    opaque_len         u32 LE   4
    opaque_data        [u8]     variable

KernelJournalV1 (exactly 209 bytes)::

    protocol_version   u32 LE   4
    kernel_version     u32 LE   4
    agent_id           bytes32  32
    agent_code_hash    bytes32  32
    constraint_set_hash bytes32 32
    input_root         bytes32  32
    execution_nonce    u64 LE   8
    input_commitment   bytes32  32
    action_commitment  bytes32  32
    execution_status   u8       1

ActionV1::

    action_type        u32 LE   4
    target             bytes32  32
    payload_len        u32 LE   4
    payload            [u8]     variable

AgentOutput::

    action_count       u32 LE   4
    for each action:
      action_len       u32 LE   4
      action           ActionV1 (variable)
"""

from __future__ import annotations

import struct
from typing import List

from .types import (
    JOURNAL_SIZE,
    KERNEL_VERSION,
    MAX_ACTION_PAYLOAD_BYTES,
    MAX_ACTIONS_PER_OUTPUT,
    MAX_AGENT_INPUT_BYTES,
    MAX_AGENT_OUTPUT_BYTES,
    MAX_SINGLE_ACTION_BYTES,
    PROTOCOL_VERSION,
    ActionV1,
    AgentOutput,
    ExecutionStatus,
    KernelInputV1,
    KernelJournalV1,
)


class CodecError(Exception):
    """Raised when encoding or decoding violates the wire format."""


# ---------------------------------------------------------------------------
# Encoding
# ---------------------------------------------------------------------------

def _encode_action(action: ActionV1) -> bytes:
    """Encode a single ActionV1 (without the outer length prefix)."""
    payload = action.payload
    if len(payload) > MAX_ACTION_PAYLOAD_BYTES:
        raise CodecError(
            f"Action payload too large: {len(payload)} bytes (limit {MAX_ACTION_PAYLOAD_BYTES})"
        )
    buf = bytearray()
    buf += struct.pack("<I", action.action_type)
    buf += action.target  # 32 bytes
    buf += struct.pack("<I", len(payload))
    buf += payload
    return bytes(buf)


def encode_output(output: AgentOutput) -> bytes:
    """Encode an ``AgentOutput`` into canonical wire bytes.

    Must match the Rust ``AgentOutput::encode()`` byte-for-byte.
    """
    n = len(output.actions)
    if n > MAX_ACTIONS_PER_OUTPUT:
        raise CodecError(f"Too many actions: {n} (limit {MAX_ACTIONS_PER_OUTPUT})")

    buf = bytearray()
    buf += struct.pack("<I", n)

    for action in output.actions:
        action_bytes = _encode_action(action)
        if len(action_bytes) > MAX_SINGLE_ACTION_BYTES:
            raise CodecError(
                f"Action too large: {len(action_bytes)} bytes (limit {MAX_SINGLE_ACTION_BYTES})"
            )
        buf += struct.pack("<I", len(action_bytes))
        buf += action_bytes

    if len(buf) > MAX_AGENT_OUTPUT_BYTES:
        raise CodecError(
            f"Output too large: {len(buf)} bytes (limit {MAX_AGENT_OUTPUT_BYTES})"
        )
    return bytes(buf)


def encode_journal(journal: KernelJournalV1) -> bytes:
    """Encode a ``KernelJournalV1`` into exactly 209 bytes.

    Must match the Rust ``KernelJournalV1::encode()`` byte-for-byte.
    """
    status_byte: int
    if journal.execution_status == ExecutionStatus.SUCCESS:
        status_byte = 0x01
    elif journal.execution_status == ExecutionStatus.FAILURE:
        status_byte = 0x02
    else:
        raise CodecError(f"Invalid execution status: {journal.execution_status}")

    buf = bytearray()
    buf += struct.pack("<I", journal.protocol_version)
    buf += struct.pack("<I", journal.kernel_version)
    buf += journal.agent_id            # 32
    buf += journal.agent_code_hash     # 32
    buf += journal.constraint_set_hash # 32
    buf += journal.input_root          # 32
    buf += struct.pack("<Q", journal.execution_nonce)
    buf += journal.input_commitment    # 32
    buf += journal.action_commitment   # 32
    buf += struct.pack("B", status_byte)

    assert len(buf) == JOURNAL_SIZE, f"Journal size mismatch: {len(buf)} != {JOURNAL_SIZE}"
    return bytes(buf)


# ---------------------------------------------------------------------------
# Decoding
# ---------------------------------------------------------------------------

def _decode_action(data: bytes) -> ActionV1:
    """Decode a single ActionV1 from raw bytes (no outer length prefix)."""
    if len(data) < 40:
        raise CodecError("Action data too short (need at least 40 bytes)")

    action_type = struct.unpack_from("<I", data, 0)[0]
    target = data[4:36]
    payload_len = struct.unpack_from("<I", data, 36)[0]

    if payload_len > MAX_ACTION_PAYLOAD_BYTES:
        raise CodecError(
            f"Action payload too large: {payload_len} bytes (limit {MAX_ACTION_PAYLOAD_BYTES})"
        )

    expected_end = 40 + payload_len
    if len(data) < expected_end:
        raise CodecError("Action data truncated")
    if len(data) != expected_end:
        raise CodecError("Trailing bytes in action data")

    payload = data[40:expected_end]
    return ActionV1(action_type=action_type, target=target, payload=payload)


def decode_output(data: bytes) -> AgentOutput:
    """Decode an ``AgentOutput`` from canonical wire bytes."""
    if len(data) > MAX_AGENT_OUTPUT_BYTES:
        raise CodecError(
            f"Output too large: {len(data)} bytes (limit {MAX_AGENT_OUTPUT_BYTES})"
        )
    if len(data) < 4:
        raise CodecError("Output data too short")

    offset = 0
    action_count = struct.unpack_from("<I", data, offset)[0]
    offset += 4

    if action_count > MAX_ACTIONS_PER_OUTPUT:
        raise CodecError(f"Too many actions: {action_count} (limit {MAX_ACTIONS_PER_OUTPUT})")

    actions: List[ActionV1] = []
    for _ in range(action_count):
        if offset + 4 > len(data):
            raise CodecError("Unexpected end of output data")
        action_len = struct.unpack_from("<I", data, offset)[0]
        offset += 4

        if action_len > MAX_SINGLE_ACTION_BYTES:
            raise CodecError(
                f"Action too large: {action_len} bytes (limit {MAX_SINGLE_ACTION_BYTES})"
            )

        action_end = offset + action_len
        if action_end > len(data):
            raise CodecError("Action data truncated")

        action = _decode_action(data[offset:action_end])
        actions.append(action)
        offset = action_end

    if offset != len(data):
        raise CodecError("Trailing bytes in output data")

    return AgentOutput(actions=actions)


def decode_input(data: bytes) -> KernelInputV1:
    """Decode a ``KernelInputV1`` from canonical wire bytes."""
    if len(data) < 148:
        raise CodecError("Input data too short (need at least 148 bytes)")

    offset = 0

    protocol_version = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    if protocol_version != PROTOCOL_VERSION:
        raise CodecError(
            f"Version mismatch: expected protocol {PROTOCOL_VERSION}, got {protocol_version}"
        )

    kernel_version = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    if kernel_version != KERNEL_VERSION:
        raise CodecError(
            f"Version mismatch: expected kernel {KERNEL_VERSION}, got {kernel_version}"
        )

    agent_id = data[offset : offset + 32]
    offset += 32
    agent_code_hash = data[offset : offset + 32]
    offset += 32
    constraint_set_hash = data[offset : offset + 32]
    offset += 32
    input_root = data[offset : offset + 32]
    offset += 32

    execution_nonce = struct.unpack_from("<Q", data, offset)[0]
    offset += 8

    opaque_len = struct.unpack_from("<I", data, offset)[0]
    offset += 4

    if opaque_len > MAX_AGENT_INPUT_BYTES:
        raise CodecError(
            f"Input too large: {opaque_len} bytes (limit {MAX_AGENT_INPUT_BYTES})"
        )

    opaque_end = offset + opaque_len
    if opaque_end > len(data):
        raise CodecError("Input data truncated")
    if opaque_end != len(data):
        raise CodecError("Trailing bytes in input data")

    opaque_agent_inputs = data[offset:opaque_end]

    return KernelInputV1(
        protocol_version=protocol_version,
        kernel_version=kernel_version,
        agent_id=agent_id,
        agent_code_hash=agent_code_hash,
        constraint_set_hash=constraint_set_hash,
        input_root=input_root,
        execution_nonce=execution_nonce,
        opaque_agent_inputs=opaque_agent_inputs,
    )


def decode_journal(data: bytes) -> KernelJournalV1:
    """Decode a ``KernelJournalV1`` from exactly 209 bytes."""
    if len(data) != JOURNAL_SIZE:
        raise CodecError(f"Journal must be exactly {JOURNAL_SIZE} bytes, got {len(data)}")

    offset = 0

    protocol_version = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    kernel_version = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    agent_id = data[offset : offset + 32]
    offset += 32
    agent_code_hash = data[offset : offset + 32]
    offset += 32
    constraint_set_hash = data[offset : offset + 32]
    offset += 32
    input_root = data[offset : offset + 32]
    offset += 32
    execution_nonce = struct.unpack_from("<Q", data, offset)[0]
    offset += 8
    input_commitment = data[offset : offset + 32]
    offset += 32
    action_commitment = data[offset : offset + 32]
    offset += 32

    status_byte = data[offset]
    offset += 1

    if status_byte == 0x01:
        execution_status = ExecutionStatus.SUCCESS
    elif status_byte == 0x02:
        execution_status = ExecutionStatus.FAILURE
    else:
        raise CodecError(f"Invalid execution status byte: 0x{status_byte:02x}")

    return KernelJournalV1(
        protocol_version=protocol_version,
        kernel_version=kernel_version,
        agent_id=agent_id,
        agent_code_hash=agent_code_hash,
        constraint_set_hash=constraint_set_hash,
        input_root=input_root,
        execution_nonce=execution_nonce,
        input_commitment=input_commitment,
        action_commitment=action_commitment,
        execution_status=execution_status,
    )
